gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientReadChannel.java

```diff
@@ -0,0 +1,568 @@
+  private final long objectSize;
+  private ContentReadChannel contentReadChannel;
+
+  private boolean open = true;
+
+  // Current position in this channel, it could be different from contentChannelCurrentPosition if
+  // position(long) method calls were made without calls to read(ByteBuffer) method.
+  private long currentPosition = 0;
+
+  public GoogleCloudStorageClientReadChannel(
+      @Nonnull Storage storage,
+      @Nonnull GoogleCloudStorageItemInfo itemInfo,
+      @Nonnull GoogleCloudStorageReadOptions readOptions)
+      throws IOException {
+    validate(itemInfo);
+    this.storage = storage;
+    this.resourceId =
+        new StorageResourceId(
+            itemInfo.getBucketName(), itemInfo.getObjectName(), itemInfo.getContentGeneration());
+    this.readOptions = readOptions;
+    this.objectSize = itemInfo.getSize();
+    this.contentReadChannel = new ContentReadChannel(readOptions, resourceId);
+  }
+
+  @Override
+  public int read(ByteBuffer dst) throws IOException {
+    throwIfNotOpen();
+
+    // Don't try to read if the buffer has no space.
+    if (dst.remaining() == 0) {
+      return 0;
+    }
+    logger.atFiner().log(
+        "Reading %d bytes at %d position from '%s'", dst.remaining(), currentPosition, resourceId);
+    if (currentPosition == objectSize) {
+      return -1;
+    }
+    return contentReadChannel.readContent(dst);
+  }
+
+  @Override
+  public int write(ByteBuffer src) throws IOException {
+    throw new UnsupportedOperationException("Cannot mutate read-only channel");
+  }
+
+  @Override
+  public long position() throws IOException {
+    return currentPosition;
+  }
+
+  /**
+   * Sets this channel's position.
+   *
+   * <p>This method will throw an exception if {@code newPosition} is greater than object size,
+   * which contradicts {@link SeekableByteChannel#position(long) SeekableByteChannel} contract.
+   * TODO(user): decide if this needs to be fixed.
+   *
+   * @param newPosition the new position, counting the number of bytes from the beginning.
+   * @return this channel instance
+   * @throws FileNotFoundException if the underlying object does not exist.
+   * @throws IOException on IO error
+   */
+  @Override
+  public SeekableByteChannel position(long newPosition) throws IOException {
+    throwIfNotOpen();
+
+    if (newPosition == currentPosition) {
+      return this;
+    }
+
+    validatePosition(newPosition);
+    logger.atFiner().log(
+        "Seek from %s to %s position for '%s'", currentPosition, newPosition, resourceId);
+    currentPosition = newPosition;
+    return this;
+  }
+
+  @Override
+  public long size() throws IOException {
+    return objectSize;
+  }
+
+  @Override
+  public SeekableByteChannel truncate(long size) throws IOException {
+    throw new UnsupportedOperationException("Cannot mutate read-only channel");
+  }
+
+  @Override
+  public boolean isOpen() {
+    return open;
+  }
+
+  @Override
+  public void close() throws IOException {
+    try {
```

Comment:
nit: Seems like it will be better if we move `try-catch-finally` block inside the `if`

Reply:
Sure.